File: CrowdCoding/src/com/crowdcoding/entities/microtasks/Review.java
package com.crowdcoding.entities.microtasks;

import static com.googlecode.objectify.ObjectifyService.ofy;

import com.crowdcoding.commands.MicrotaskCommand;
import com.crowdcoding.commands.WorkerCommand;
import com.crowdcoding.dto.DTO;
import com.crowdcoding.dto.ReviewDTO;
import com.crowdcoding.dto.firebase.MicrotaskInFirebase;
import com.crowdcoding.dto.firebase.NewsItemInFirebase;
import com.crowdcoding.dto.firebase.NotificationInFirebase;
import com.crowdcoding.dto.firebase.ReviewInFirebase;
import com.crowdcoding.entities.Artifact;
import com.crowdcoding.entities.Function;
import com.crowdcoding.entities.Project;
import com.crowdcoding.history.HistoryLog;
import com.crowdcoding.history.MicrotaskAccepted;
import com.crowdcoding.history.MicrotaskReissued;
import com.crowdcoding.history.MicrotaskRejected;
import com.crowdcoding.history.MicrotaskSpawned;
import com.crowdcoding.util.FirebaseService;
import com.googlecode.objectify.Key;
import com.googlecode.objectify.Ref;
import com.googlecode.objectify.annotation.Subclass;
import com.googlecode.objectify.annotation.Load;
import com.googlecode.objectify.annotation.Parent;

@Subclass(index=true)
public class Review extends Microtask
{
	@Parent @Load private Ref<Artifact> artifact;
	private Key<Microtask> microtaskKeyUnderReview;
	private String workerOfReviewedWork;
	private String initiallySubmittedDTO;	// initially submitted DTO in string format

	// Default constructor for deserialization
	private Review()
	{
	}

	// Constructor for initial construction
	public Review(Key<Microtask> microtaskKeyUnderReview, String initiallySubmittedDTO, String workerOfReviewedWork,
			String projectId)
	{
		super(projectId);
		this.submitValue = 5;
		this.microtaskKeyUnderReview = microtaskKeyUnderReview;
		this.initiallySubmittedDTO = initiallySubmittedDTO;
		this.workerOfReviewedWork = workerOfReviewedWork;

		Microtask microtaskUnderReview = ofy().load().key(microtaskKeyUnderReview).now();
		this.artifact = (Ref<Artifact>) Ref.create((Key<Artifact>) microtaskUnderReview.getOwningArtifact().getKey());

		ofy().save().entity(this).now();
		FirebaseService.writeMicrotaskCreated(new ReviewInFirebase(
				id,
				this.microtaskTitle(),
				this.microtaskName(),
				this.artifact.get().getName(),
				this.artifact.get().getID(),
				false,
				false,
				submitValue,
				 microtaskKeyUnderReview
				),
				Microtask.keyToString(this.getKey()),
				projectId);

		HistoryLog.Init(projectId).addEvent(new MicrotaskSpawned(this));
	}

    public Microtask copy(String projectId)
    {
    	return new Review(microtaskKeyUnderReview, this.initiallySubmittedDTO, this.workerOfReviewedWork, projectId);
    }

	protected void doSubmitWork(DTO dto, String workerID, String projectId)
	{

		ReviewDTO reviewDTO = (ReviewDTO) dto;

		Microtask submittedMicrotask = Microtask.find(microtaskKeyUnderReview).now();

		// Write the review to firebase
		FirebaseService.writeReview(reviewDTO, this.id, Microtask.keyToString(submittedMicrotask.getKey()) , projectId);

		// set default award points to 0
		int points = 0;


		int awardedPoints;
		String reviewResult;

		//reject not used any more for now...
        /*if( reviewDTO.qualityScore < 3 ) {

			// reissue microtask
        	System.out.println("--> REVIEW mtask "+Project.MicrotaskKeyToString( submittedMicrotask.getKey() )+" rejected");
			MicrotaskCommand.rejectAndReissueMicrotask(microtaskKeyUnderReview, workerOfReviewedWork);
			awardedPoint = 0;
			reviewResult = "rejected";


			project.historyLog().beginEvent(new MicrotaskRejected(submittedMicrotask,workerID));
			project.historyLog().endEvent();

		} else*/

		/*the awarded point is proportional to the square of the total points
			points : 10
			5 stars ->10
			4 stars ->6
			3 stars ->3
			2 stars ->1
			1 stars ->0
		*/
		awardedPoints = (int) Math.round( submittedMicrotask.submitValue * Math.pow((reviewDTO.qualityScore /5.0), 1.5));
		NotificationInFirebase notification = null;
		if ( reviewDTO.qualityScore < 4) {

			// reissue microtask
			System.out.println("--> REVIEW mtask "+submittedMicrotask.getKey().toString()+" reissued");
			reviewResult = "reissued";
			if(reviewDTO.fromDisputedMicrotask)
				MicrotaskCommand.rejectMicrotask(microtaskKeyUnderReview, workerOfReviewedWork, awardedPoints);
			else
				MicrotaskCommand.reviseMicrotask(microtaskKeyUnderReview, initiallySubmittedDTO, reviewDTO.reviewText, workerOfReviewedWork, awardedPoints);

			HistoryLog.Init(projectId).addEvent(new MicrotaskReissued(submittedMicrotask,workerID));

			notification = new NotificationInFirebase(
					"task.reissued",
					"{ \"microtaskId\": \""+submittedMicrotask.getID() + "\", \"microtaskType\": \""+submittedMicrotask.microtaskName() + "\", \"artifactName\": \""+submittedMicrotask.getOwningArtifact().getName() + "\"}"
			);

		} else {

			// accept microtask
        	System.out.println("--> REVIEW mtask "+submittedMicrotask.getKey().toString()+" accepted");
			reviewResult ="accepted";
			MicrotaskCommand.submit(microtaskKeyUnderReview, initiallySubmittedDTO, workerOfReviewedWork, awardedPoints);


			HistoryLog.Init(projectId).addEvent(new MicrotaskAccepted(submittedMicrotask,workerID));
			
			notification = new NotificationInFirebase(
					"task.accepted",
					"{ \"microtaskId\": \""+submittedMicrotask.getID() + "\", \"microtaskType\": \""+submittedMicrotask.microtaskName() + "\", \"artifactName\": \""+submittedMicrotask.getOwningArtifact().getName() + "\"}"
			);

		}


		// send feedback
    	FirebaseService.postToNewsfeed(workerOfReviewedWork, (
    		new NewsItemInFirebase(
    			awardedPoints,
    			submittedMicrotask.submitValue,
    			submittedMicrotask.microtaskName(),
				"WorkReviewed",
				Microtask.keyToString(submittedMicrotask.getKey()),
				reviewDTO.qualityScore)
	    	).json(),
	    	Microtask.keyToString(submittedMicrotask.getKey()),
    		projectId
	    );
    	// send notification
		FirebaseService.writeWorkerNotification(
				notification,
				workerOfReviewedWork, 
				projectId
		);



		//FirebaseService.setPoints(workerID, workerOfReviewedWork,  this.submitValue, project);
    	FirebaseService.postToNewsfeed(workerID, (
    		new NewsItemInFirebase(
    			this.submitValue,
    			this.submitValue,
    			this.microtaskName(),
    			"SubmittedReview",
    			Microtask.keyToString(this.getKey()),
    			-1 // differentiate the reviews from the 0 score tasks
	    	).json()),
	    	Microtask.keyToString(this.getKey()),
			projectId
		);

		// increase the stats counter
		WorkerCommand.increaseStat(workerID, "reviews",1);

	}

    public Key<Microtask> getKey()
	{
		return Key.create( artifact.getKey(), Microtask.class, this.id );
	}

	protected Class getDTOClass()
	{
		return ReviewDTO.class;
	}

	public Key<Microtask> getMicrotaskIDUnderReview()
	{
		return microtaskKeyUnderReview;
	}

	public String getUIURL()
	{
		return "/html/microtasks/review.jsp";
	}


	public Artifact getOwningArtifact()
	{
		Artifact owning;
		try {
			return artifact.safe();
		} catch ( Exception e ){
			ofy().load().ref(this.artifact);
			return artifact.get();
		}
	}

	public String microtaskTitle()
	{
		return "Review work";
	}

	public String microtaskDescription()
	{
		return "review a submitted microtask";
	}

}

File: CrowdCoding/war/node/entities/microtasks/Review.js
package com.crowdcoding.entities.microtasks;

import static com.googlecode.objectify.ObjectifyService.ofy;

import com.crowdcoding.commands.WorkerCommand;
import com.crowdcoding.dto.DTO;
import com.crowdcoding.dto.ajax.microtask.submission.ImplementBehaviorDTO;
import com.crowdcoding.dto.firebase.microtasks.ImplementBehaviorInFirebase;
import com.crowdcoding.entities.artifacts.Artifact;
import com.crowdcoding.entities.artifacts.Function;
import com.crowdcoding.history.HistoryLog;
import com.crowdcoding.history.MicrotaskSpawned;
import com.crowdcoding.util.FirebaseService;
import com.googlecode.objectify.Key;
import com.googlecode.objectify.Ref;
import com.googlecode.objectify.annotation.Subclass;
import com.googlecode.objectify.annotation.Load;
import com.googlecode.objectify.annotation.Parent;

var http = require('http');

var server = http.createServer(function(req, res) {
  res.writeHead(200);
  res.end('Hello Http');
});
server.listen(8080);

@Subclass(index=true)
class ImplementBehavior extends Microtask
{
	public enum PromptType { WRITE, CALLEE_CHANGE, REMOVE_CALLEE, CORRECT };
	@Parent @Load var function;
	var promptType;

	var testId;					//Only defined for WRITE

	var oldCalleeVersion;			//Only defined for CALLEE_CHANGED

	var disputeText;				// Only defined for CORRECT and REMOVE_CALLEE
	var calleeId;					// Only defined for REMOVE_CALLEE and CALLEE_CHANGED

	var disputeId;				//id of the artifact that disputed this function

	// Default constructor for deserialization
	constructor ImplementBehavior()
	{
	}

	// Initialization constructor for a WRITE write function. Microtask is not ready.
	constructor(function, testId, projectId)
	{
		super(projectId, function.getId());

		this.promptType = PromptType.WRITE;

		this.testId		= testId;

		implementBehavior(function, projectId);
	}

	// Initialization constructor for a DESCRIPTION_CHANGE write function. Microtask is not ready.
	constructor(function, calleeId,
			int oldCalleeVersion, String projectId)
	{
		super(projectId,function.getId());
		this.promptType = PromptType.CALLEE_CHANGE;

		this.calleeId = calleeId;
		this.oldCalleeVersion = oldCalleeVersion;

		implementBehavior(function, projectId);
	}

	// Initialization constructor for a RE_EDIT write function. Microtask is not ready.
	constructor(function, disputeText, calleeId, projectId)
	{
		super(projectId,function.getId());
		this.promptType = PromptType.REMOVE_CALLEE;

		this.disputeText = disputeText;
		this.calleeId = calleeId;
		implementBehavior(function, projectId);
	}
	constructor(function, disputeText, projectId)
	{
		super(projectId,function.getId());
		this.promptType = PromptType.CORRECT;

		this.disputeText = disputeText;

		implementBehavior(function, projectId);
	}


	function copy(projectId)
	{
		switch (promptType) {
			case WRITE:
				return new ImplementBehavior( (Function) getOwningArtifact() , testId, projectId);
			case CALLEE_CHANGE:
				return new ImplementBehavior( (Function) getOwningArtifact() , calleeId, oldCalleeVersion, projectId);
			case REMOVE_CALLEE:
				return new ImplementBehavior( (Function) getOwningArtifact() , this.disputeText, disputeId, projectId);
			case CORRECT:
				return new ImplementBehavior((Function) getOwningArtifact(), this.disputeText, projectId);
	
			default:
				return null;
		}
	}

	function getKey()
	{
		return Key.create( function.getKey(), Microtask.class, this.id );
	}


	function implementBehavior(function, projectId)
	{
		this.function = function.getRef();
		ofy().load().ref(this.function);
		ofy().save().entity(this).now();
		FirebaseService.writeMicrotaskCreated(new ImplementBehaviorInFirebase(
				id,
				this.microtaskTitle(),
				this.microtaskName(),
				function.getName(),
				function.getId(),
				submitValue,
				function.getId(),
				this.promptType.name(),
				this.testId,
				this.calleeId,
				this.oldCalleeVersion,
				this.disputeText),
				Microtask.keyToString(this.getKey()),
				projectId);


		HistoryLog.Init(projectId).addEvent(new MicrotaskSpawned(this));
	}

	function doSubmitWork(dto, workerID)
	{
		function.get().implementBehaviorCompleted((ImplementBehaviorDTO) dto, disputeId , projectId);
//		WorkerCommand.awardPoints(workerID, this.submitValue);
		// increase the stats counter
		WorkerCommand.increaseStat(workerID, "functions",1);
	}

	function getPromptType()
	{
		return promptType;
	}

	function getDTOClass()
	{
		return ImplementBehaviorDTO.class;
	}

	function getFunction()
	{
		return function.getValue();
	}


	function getOwningArtifact()
	{
		Artifact owning;
		try {
			return function.safe();
		} catch ( Exception e ){
			ofy().load().ref(this.function);
			return function.get();
		}
	}

	function microtaskTitle()
	{
		return "Implement function behavior";
	}

	function microtaskDescription()
	{
		return "implement function behaviorn";
	}


}
